packages/zowe-explorer-api/src/profiles/AuthHandler.ts

```diff
@@ -0,0 +1,193 @@
+    errorCorrelation?: CorrelatedError;
+}
+
+type ProfileLike = string | imperative.IProfileLoaded;
+export class AuthHandler {
+    private static profileLocks: Map<string, Mutex> = new Map();
+
+    /**
+     * Function that checks whether a profile is using token based authentication
+     * @param {string[]} secureProfileProps Secure properties for the service profile
+     * @param {string[]} baseSecureProfileProps Base profile's secure properties (optional)
+     * @returns {Promise<boolean>} a boolean representing whether token based auth is being used or not
+     */
+    public static isUsingTokenAuth(secureProfileProps: string[], baseSecureProfileProps?: string[]): boolean {
+        const profileUsesBasicAuth = secureProfileProps.includes("user") && secureProfileProps.includes("password");
+        if (secureProfileProps.includes("tokenValue")) {
+            return secureProfileProps.includes("tokenValue") && !profileUsesBasicAuth;
+        }
+        return baseSecureProfileProps?.includes("tokenValue") && !profileUsesBasicAuth;
+    }
+
+    /**
+     * Unlocks the given profile so it can be used again.
+     * @param profile {ProfileLike} The profile (name or {@link imperative.IProfileLoaded} object) to unlock
+     * @param refreshResources {boolean} Whether to refresh high-priority resources (active editor & virtual workspace) after unlocking
+     */
+    public static unlockProfile(profile: ProfileLike, refreshResources?: boolean): void {
+        const profileName = typeof profile === "string" ? profile : profile.name;
+        const mutex = this.profileLocks.get(profileName);
+        // If a mutex doesn't exist for this profile or the mutex is no longer locked, return
+        if (mutex == null || !mutex.isLocked()) {
+            return;
+        }
+
+        mutex.release();
+        if (refreshResources) {
+            // TODO: Log errors using ZoweLogger once available in ZE API
+            // refresh an active, unsaved editor if it uses the profile
+            FileManagement.reloadActiveEditorForProfile(profileName)
+                // eslint-disable-next-line no-console
+                .catch((err) => err instanceof Error && console.error(err.message));
+
+            // refresh virtual workspaces for the profile
+            FileManagement.reloadWorkspacesForProfile(profileName)
+                // eslint-disable-next-line no-console
+                .catch((err) => err instanceof Error && console.error(err.message));
+        }
+    }
+
+    private static async updateTreeProvidersWithProfile(profile: imperative.IProfileLoaded): Promise<void> {
+        // TODO: If we can access extender tree providers (e.g. CICS), it would help to propagate profile updates here.
+        // For now we will propagate profile changes to core providers (Data Sets, USS, Jobs)
+        const treeProviders = (await commands.executeCommand("zowe.getTreeProviders")) as any;
```

Comment:
This is a clever way to access the tree providers for refreshing profile data across all of them 😋 

It feels a bit weird to me to be using a command for this. Curious if an event-driven approach would be more extensible? Perhaps the ZE API could fire a `profileUpdated` event that each of the tree providers handles.

Reply:
I agree that an event would be more extensible, this should be resolved in e354184, will address any missing patch coverage once I'm back on 1/3 😋